packages/config/src/headers.js

```diff
@@ -0,0 +1,52 @@
+'use strict'
+
+const { resolve } = require('path')
+
+// @todo: use `util.isDeepStrictEqual()` after dropping support for Node 8
+const fastDeepEqual = require('fast-deep-equal')
+const { parseFileHeaders, mergeHeaders, normalizeHeaders } = require('netlify-headers-parser')
+
+const { warnHeadersParsing } = require('./log/messages')
+
+// Add `config.headers`
+const addHeaders = async function (config, logs) {
+  const headersPath = resolve(config.build.publish, HEADERS_FILENAME)
+  try {
+    const configWithHeaders = await addConfigHeaders(config, headersPath)
+    return { config: configWithHeaders, headersPath }
+    // @todo remove this failsafe once the code is stable
+  } catch (error) {
+    warnHeadersParsing(logs, error.message)
+    return { config: { ...config, headers: [] }, headersPath }
+  }
+}
+
+const HEADERS_FILENAME = '_headers'
+
+const addConfigHeaders = async function ({ headers: configHeaders = [], ...config }, headersPath) {
+  const normalizedConfigHeaders = normalizeHeaders(configHeaders)
+  const normalizedFileHeaders = await getFileHeaders(headersPath, normalizedConfigHeaders)
+  const headers = mergeHeaders({ fileHeaders: normalizedFileHeaders, configHeaders: normalizedConfigHeaders })
+  return { ...config, headers }
+}
+
+const getFileHeaders = async function (headersPath, normalizedConfigHeaders) {
+  const fileHeaders = await parseFileHeaders(headersPath)
+  const normalizedFileHeaders = normalizeHeaders(fileHeaders)
+  return hasMergedFileHeaders(normalizedFileHeaders, normalizedConfigHeaders) ? [] : normalizedFileHeaders
+}
+
+// `configHeaders` might contain the content of `_headers` already.
+// This happens when a plugin appends to `netlifyConfig.headers` which already
+// contains `_headers` rules and is transformed to an `inlineConfig` object,
+// which is itself parsed as `configHeaders`. In that case, we do not want
+// duplicate so we ignore `fileHeaders`.
+// The current logic works with the case where `_headers` is added both before
+// and after modifying `netlifyConfig.headers`.
+const hasMergedFileHeaders = function (normalizedFileHeaders, normalizedConfigHeaders) {
+  return normalizedConfigHeaders.some((configHeader) =>
+    normalizedFileHeaders.some((fileHeader) => fastDeepEqual(configHeader, fileHeader)),
+  )
+}
```

Comment:
Probably not following the logic correctly, so apologies if what I'm asking is total nonsense 😄 but given we're merging these headers afterwards this is a mere performance enhancement correct? We're just validating that the `configHeaders` already contain some `_headers` and avoid that merge logic? Or we're not performing a deep merge and as such this is required in order to avoid dupicates? Also, on that note, we're checking the presence of `some` common header in both file and config and if we find one, we assume there's nothing to merge, couldn't there be a case where we have one common header, but still different headers that need/should be merged in both arrays?

Reply:
Thanks for the overview!

> Headers from _headers and netlify.toml are concatenated, not deep merged (this is how the backend handles it).

This pretty much covers 99% of the questions 👍

> Good point. Yes, that's possible in principle, although this feels like an edge case. I am leaning towards waiting to see if this turns out to be a problem before implementing a solution since that might not be very simple to implement, what do you think?
>
> Note: the reason for some() instead of every() is because some of the duplicate headers might have been modified by plugins, i.e. would not be identical anymore.

Sounds good to me 👍

Reply:
Follow-up PR about this in https://github.com/netlify/build/pull/3424